### DISCUSSION TITLE
Add support for TVOS platform

### DISCUSSION THREAD

**Discussion by nriccar** [invertase/react-native-firebase#6526] [Ideas]
Hello,

I would be awesome to have TVOS targets available.

[react-native-tvos ](https://github.com/react-native-tvos/react-native-tvos) + firebase would be great!

**Comment by mikehardy** [comment]
Hi there! PRs happily accepted. It's probably not that hard either - it is mostly `#if` statements for tvOS target to comment out things that don't work for the platform / adding `@available`  for APIs

some of the work has already been done
You basically just start building it and fixing the errors one by one and it's likely just the work of a couple hours and you have a PR

the work results in hits like this, but for available target tvOS / compiler flag TARGET_OS_TV


```
mike@bistromath:~/work/Invertase/react-native-firebase (main) % grep -r available packages/**/ios

packages/app/ios/RNFBApp/RNFBUtilsModule.m:      if (@available(macOS 11, iOS 12, *)) {
packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m:      code = @"unavailable";
packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m:      message = @"The service is unavailable.";
packages/firestore/ios/RNFBFirestore/RNFBFirestoreCommon.m:    case FIRFirestoreErrorCodeUnavailable:
packages/firestore/ios/RNFBFirestore/RNFBFirestoreCommon.m:      code = @"unavailable";
packages/firestore/ios/RNFBFirestore/RNFBFirestoreCommon.m:      message = @"The service is currently unavailable. This is a most likely a transient "
packages/functions/ios/RNFBFunctions/RNFBFunctionsModule.m:    case FIRFunctionsErrorCodeUnavailable:
packages/messaging/ios/RNFBMessaging/RNFBMessagingSerializer.m:    if (apsDict[@"content-available"] != nil) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingSerializer.m:      message[@"contentAvailable"] = @([RCTConvert BOOL:apsDict[@"content-available"]]);
packages/messaging/ios/RNFBMessaging/RNFBMessaging+UNUserNotificationCenter.m:      if (@available(iOS 14, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessaging+UNUserNotificationCenter.m:      if (@available(iOS 14, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessaging+UNUserNotificationCenter.m:    if (@available(iOS 12.0, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:                           @"code" : @"unavailable-in-extension",
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:  if (@available(iOS 10.0, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:      if (@available(iOS 12.0, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:      if (@available(iOS 12.0, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:      if (@available(iOS 13.0, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:      if (@available(iOS 12, macOS 10.14, macCatalyst 13, tvOS 12, watchOS 5, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:  if (@available(iOS 10.0, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:  if (@available(iOS 10.0, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:          if (@available(iOS 12.0, *)) {
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:          if (@available(iOS 14.0, macCatalyst 14.0, *)) {
```

```
mike@bistromath:~/work/Invertase/react-native-firebase (main) % grep -ri '#if' packages/**/ios
packages/app/ios/RNFBApp/RNFBUtilsModule.m:#if (!TARGET_OS_MACCATALYST)
packages/app/ios/RNFBApp/RNFBAppModule.m:#if __has_include(<FirebaseCore/FIRAppInternal.h>)
packages/app/ios/RNFBApp/RNFBAppModule.m:#ifdef REGISTER_LIB
packages/app/ios/RNFBApp/RNFBSharedUtils.h:#ifndef RNFBSharedUtils_h
packages/app/ios/RNFBApp/RNFBSharedUtils.h:#ifdef DEBUG
packages/crashlytics/ios/RNFBCrashlytics/RNFBCrashlyticsInitProvider.m:#ifdef DEBUG
packages/messaging/ios/RNFBMessaging/RNFBMessaging+NSNotificationCenter.m:#if !(TARGET_IPHONE_SIMULATOR)
packages/messaging/ios/RNFBMessaging/RNFBMessaging+NSNotificationCenter.m:#if !(TARGET_IPHONE_SIMULATOR)
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:#if !(TARGET_IPHONE_SIMULATOR)
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:#if !(TARGET_IPHONE_SIMULATOR)
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 130000
packages/messaging/ios/RNFBMessaging/RNFBMessagingModule.m:#if TARGET_IPHONE_SIMULATOR
packages/messaging/ios/RNFBMessaging/RNFBMessaging+AppDelegate.m:#ifdef DEBUG
packages/messaging/ios/RNFBMessaging/RNFBMessaging+AppDelegate.m:#if __has_include(<FirebaseAuth/FirebaseAuth.h>)

```

**Reply by vladinko0** [reply]
Where can be found your PR?

**Reply by mikehardy** [reply]
No one has proposed a PR. @vladinko0 are you interested in proposing one?

**Reply by vladinko0** [reply]
Maybe I could help, but I don't know how to do it. I need more information even how to start, how to check, run etc...

**Reply by mikehardy** [reply]
This may not be a productive task for you then. You'll need a tvOS simulator on your mac, and you'll need to get react-native-tvos working, and install react-native-firebase in it. Then you'll need to try to compile a test app for the tvOS target and fix compile errors as you go - making reasonable decisions between "can just #ifdef this out for tvOS builds" vs "need to handle this with special logic somehow"

**Reply by vladinko0** [reply]
I have tvOS simulator on my mac and succesfuly running react-native based apps.

**Reply by vladinko0** [reply]
In fact I just need it for Analytics.

**Reply by mikehardy** [reply]
If you propose a PR that makes the app and analytics modules work for tvOS, I'd be happy to merge it.

**Reply by vladinko0** [reply]
I would be happy to do it. Can you please guide me with the first steps?

**Reply by vladinko0** [reply]
I have cloned https://github.com/invertase/react-native-firebase/

**Reply by mikehardy** [reply]
Open the e2e app in Xcode and attempt to build it targeted for your tvOS simulator - we use our e2e app for development, and it has a guide here: https://github.com/invertase/react-native-firebase/blob/main/tests/README.md

**Reply by vladinko0** [reply]
I opened react-native-firebase/tests/ios/testing.xcworkspace in Xcode, added AppleTv destination and launched and Error is:
`'React/RCTBridgeDelegate.h' file not found `

**Reply by vladinko0** [reply]
From mentioned guide, when I run yarn `tests:ios:build` I get:

> yarn run v1.22.18
> $ cd tests && yarn detox build --configuration ios.sim.debug
> $ /Users/vladimirskoupil/Projekty/react-native-firebase/tests/node_modules/.bin/detox build --configuration ios.sim.debug
> 19:49:12.969 detox[54566] INFO:  [build.js] set -o pipefail && xcodebuild VALID_ARCHS="`uname -m`"  CC=clang CPLUSPLUS=clang++ LD=clang LDPLUSPLUS=clang++ -workspace ios/testing.xcworkspace -scheme testing -configuration Debug -sdk iphonesimulator -derivedDataPath ios/build | xcbeautify
> /bin/sh: xcbeautify: command not found
> --- xcodebuild: WARNING: Using the first of multiple matching destinations:
> { platform:iOS Simulator, id:dvtdevice-DVTiOSDeviceSimulatorPlaceholder-iphonesimulator:placeholder, name:Any iOS Simulator Device }
> { platform:iOS Simulator, id:72592D38-0C23-4A8F-AEBB-81B7F346E596, OS:14.3, name:iPhone 8 }
> { platform:iOS Simulator, id:23B67596-272A-47E6-A71A-9C715453CFE9, OS:15.2, name:iPhone 8 }
> { platform:iOS Simulator, id:055667B4-4F7A-4BC0-9F70-AA1F5C156DA6, OS:14.3, name:iPhone 8 Plus }
> { platform:iOS Simulator, id:14426D38-AEC4-4A6B-A94C-B4BFC9F06B03, OS:15.2, name:iPhone 8 Plus }
> { platform:iOS Simulator, id:489276F3-2F05-4F9E-85AA-5157934C4EEE, OS:14.3, name:iPhone 11 }
> { platform:iOS Simulator, id:07EA0DEB-9D33-4538-98DB-561EBF50405A, OS:15.2, name:iPhone 11 }
> { platform:iOS Simulator, id:E04CAEF2-8F03-4E6A-B507-F964D210CFE9, OS:15.2, name:iPhone 11 15.2 }
> { platform:iOS Simulator, id:CCAF8708-9532-480B-9B7F-D924674F21E1, OS:14.3, name:iPhone 11 Pro }
> { platform:iOS Simulator, id:9875222F-0B5D-4EE6-9EB9-0E409FFC8BD5, OS:15.2, name:iPhone 11 Pro }
> { platform:iOS Simulator, id:1143319A-28C8-4BBC-8F4E-08F6ACC03B27, OS:14.3, name:iPhone 11 Pro Max }
> { platform:iOS Simulator, id:114A2D6E-24DF-40C8-BF50-F14BC528E328, OS:15.2, name:iPhone 11 Pro Max }
> { platform:iOS Simulator, id:52644113-7679-4797-89E4-EF771649A638, OS:14.3, name:iPhone 12 }
> { platform:iOS Simulator, id:40DD9575-FA27-4F45-A2A6-6116095D88A7, OS:14.3, name:iPhone 12 Pro }
> { platform:iOS Simulator, id:A99CEA4F-B01E-4D0C-A9FE-FF6BDB8A45D8, OS:15.2, name:iPhone 12 Pro }
> { platform:iOS Simulator, id:87678E58-810C-4486-9FD2-BF3886070A1F, OS:14.3, name:iPhone 12 Pro Max }
> { platform:iOS Simulator, id:773A3C9A-B74F-48C2-BCBF-899433AE95DB, OS:15.2, name:iPhone 12 Pro Max }
> { platform:iOS Simulator, id:A0B5AE00-C40F-40A7-9FBC-8BEEC911C390, OS:15.2, name:iPhone 12 Tim }
> { platform:iOS Simulator, id:0673280B-F12F-444B-AD3D-39F470CE9777, OS:14.3, name:iPhone 12 mini }
> { platform:iOS Simulator, id:2F3797FE-C4A9-46AE-B96E-C80D9A935095, OS:15.2, name:iPhone 12 mini }
> { platform:iOS Simulator, id:68B732A6-15E2-49DA-85C4-67A1E38CE2A8, OS:15.2, name:iPhone 13 }
> { platform:iOS Simulator, id:6AE55CCA-F51C-402F-9940-BC0355071C10, OS:15.2, name:iPhone 13 Pro }
> { platform:iOS Simulator, id:B4FC1DD5-39C1-4D49-8853-572D4B4F6734, OS:15.2, name:iPhone 13 Pro Max }
> { platform:iOS Simulator, id:807FC862-839C-4D87-9D1B-DD292EC93B83, OS:15.2, name:iPhone 13 mini }
> { platform:iOS Simulator, id:62FA867D-63E5-4127-B7A8-F9816E11DAE3, OS:16.1, name:iPhone 14 }
> { platform:iOS Simulator, id:034B866D-B917-4AFD-93BB-CE6F5BF273E1, OS:16.1, name:iPhone 14 Plus }
> { platform:iOS Simulator, id:91B698D2-02A3-4DD9-AC8B-EBA2E4A825D7, OS:16.1, name:iPhone 14 Pro }
> { platform:iOS Simulator, id:5E50A2D2-FE2A-4CA0-94BD-E1E706FC61DD, OS:16.1, name:iPhone 14 Pro Max }
> { platform:iOS Simulator, id:DEAF16EE-44FC-4D8C-8039-FBE0F991FF62, OS:14.3, name:iPhone SE (2nd generation) }
> { platform:iOS Simulator, id:E4CAC463-09DE-417B-B1AA-A0844F41D1AE, OS:15.2, name:iPhone SE (2nd generation) }
> { platform:iOS Simulator, id:EC708BC4-0142-4A57-8F82-25B74160CFB5, OS:16.1, name:iPhone SE (3rd generation) }
> { platform:iOS Simulator, id:6C3354A9-4DBE-4DBD-8AE3-F143406A69E3, OS:14.3, name:iPod touch (7th generation) }
> { platform:iOS Simulator, id:4F1386DC-B6F4-4478-8036-6C8247F4C839, OS:15.2, name:iPod touch (7th generation) }
> { platform:iOS, id:dvtdevice-DVTiPhonePlaceholder-iphoneos:placeholder, name:Any iOS Device }
> { platform:tvOS, id:dvtdevice-DVTiOSDevicePlaceholder-appletvos:placeholder, name:Any tvOS Device }
> { platform:tvOS Simulator, id:dvtdevice-DVTiOSDeviceSimulatorPlaceholder-appletvsimulator:placeholder, name:Any tvOS Simulator Device }
> { platform:tvOS Simulator, id:E0ED029A-BBDA-4AF0-A093-624F36E911E2, OS:16.1, name:Apple TV }
> { platform:tvOS Simulator, id:179ABAB6-280B-45E2-9743-7922121EB83E, OS:16.1, name:Apple TV 4K (3rd generation) }
> { platform:tvOS Simulator, id:2E09AAFA-E52D-49B0-8A81-A3F20FA6F3CC, OS:16.1, name:Apple TV 4K (3rd generation) (at 1080p) }
> { platform:iOS Simulator, id:863681D9-73BF-4452-A440-6DB1F315E925, OS:14.3, name:iPad (8th generation) }
> { platform:iOS Simulator, id:3D294039-9CC9-4FD1-8C92-52F84A7FB180, OS:15.2, name:iPad (9th generation) }
> { platform:iOS Simulator, id:FBDABBE8-6BB7-45DB-9E3A-87FFE1F6561D, OS:16.1, name:iPad (10th generation) }
> { platform:iOS Simulator, id:D8FD854F-5422-4737-A8C9-A5104C4134CE, OS:16.1, name:iPad Air (3rd generation) }
> { platform:iOS Simulator, id:087164AD-29C9-4F71-AA17-727871FF1167, OS:14.3, name:iPad Air (4th generation) }
> { platform:iOS Simulator, id:D42728AD-97C4-4315-BCC8-00E30D882240, OS:15.2, name:iPad Air (4th generation) }
> { platform:iOS Simulator, id:331E8B6A-8B54-4400-A416-D3047B5599EA, OS:16.1, name:iPad Air (5th generation) }
> { platform:iOS Simulator, id:0A983C2F-3A89-4B26-A07C-C45442BACA4A, OS:14.3, name:iPad Pro (9.7-inch) }
> { platform:iOS Simulator, id:B29A4B7A-4A79-4F0C-8508-8BA121832141, OS:15.2, name:iPad Pro (9.7-inch) }
> { platform:iOS Simulator, id:745F6E70-626B-4899-A1AA-5ED54C694836, OS:14.3, name:iPad Pro (11-inch) (2nd generation) }
> { platform:iOS Simulator, id:85302E7C-37C2-4A4C-980E-17E6CB10193B, OS:15.2, name:iPad Pro (11-inch) (3rd generation) }
> { platform:iOS Simulator, id:4733DFCF-995C-4651-A4C1-F56F16B975AE, OS:16.1, name:iPad Pro (11-inch) (4th generation) }
> { platform:iOS Simulator, id:3ECE3B41-0A1A-4ADE-9CB0-599898879839, OS:14.3, name:iPad Pro (12.9-inch) (4th generation) }
> { platform:iOS Simulator, id:717CD3EB-C8C9-4CB6-BEEC-F1DE49EF62F1, OS:15.2, name:iPad Pro (12.9-inch) (5th generation) }
> { platform:iOS Simulator, id:D0E30F9E-76D3-4F3E-9FF7-983AEC79A1C9, OS:16.1, name:iPad Pro (12.9-inch) (6th generation) }
> { platform:iOS Simulator, id:E7227BA3-2B16-4FB8-ACF8-79CDC7CE0E24, OS:15.2, name:iPad mini (6th generation) }
> { platform:iOS Simulator, id:8AA054E0-8699-4DF5-9DAF-E1A27BB9BAA4, OS:16.1, name:iPad mini (6th generation) }

and it stayed without any progress...

**Comment by vladinko0** [comment]
Somehow solved but then with `yarn tests:ios:test`:

> yarn run v1.22.18
> $ cd tests && SIMCTL_CHILD_GULGeneratedClassDisposeDisabled=1 yarn detox test --configuration ios.sim.debug --loglevel warn
> $ /Users/vladimirskoupil/Projekty/react-native-firebase/tests/node_modules/.bin/detox test --configuration ios.sim.debug --loglevel warn
> 20:04:46.181 detox[56425] ERROR: at node_modules/mocha/lib/cli/run.js:372:13 
>  
> Error: Cannot find module '../../../.github/workflows/scripts/functions/lib/sample-data'
> Require stack:

**Reply by mikehardy** [reply]
Have you run `yarn tests:emulator:start` yet? The sample data is in typescript format but when the firebase emulator starts, the functions should be compiled from src into that lib path 🤔

**Reply by vladinko0** [reply]
It is running now:
![Screenshot 2023-03-24 at 07 59 28](https://user-images.githubusercontent.com/16008547/227448345-41d76267-0253-472d-94c7-a92f8955a317.png)

How should I change target to tvOS?

**Reply by vladinko0** [reply]
I changed in `tests/package.json` `detox.configurations.ios.sim.debug.device.type` and in `detox.configurations.ios.sim.release.device.type` from `iPhone 14` to `Apple TV`.

Then I run `yarn tests:ios:test `and I have errors:

> 08:04:06.851 detox[4197] ERROR: [EXEC_FAIL, #8] "/usr/bin/xcrun simctl install E0ED029A-BBDA-4AF0-A093-624F36E911E2 "/Users/vladimirskoupil/Projekty/react-native-firebase/tests/ios/build/Build/Products/Debug-iphonesimulator/testing.app"" failed with error = ChildProcessError: Command failed: /usr/bin/xcrun simctl install E0ED029A-BBDA-4AF0-A093-624F36E911E2 "/Users/vladimirskoupil/Projekty/react-native-firebase/tests/ios/build/Build/Products/Debug-iphonesimulator/testing.app"
> An error was encountered processing the command (domain=IXUserPresentableErrorDomain, code=6):
> Unable To Install “testing”
> This app is not made for this device.
> This app was not built to support this device family; app is compatible with (
>     1
> ) but this device supports (
>     3
> )
> Underlying error (domain=MIInstallerErrorDomain, code=10):
>         This app was not built to support this device family; app is compatible with (
>     1
> ) but this device supports (
>     3
> )
>  `/usr/bin/xcrun simctl install E0ED029A-BBDA-4AF0-A093-624F36E911E2 "/Users/vladimirskoupil/Projekty/react-native-firebase/tests/ios/build/Build/Products/Debug-iphonesimulator/testing.app"` (exited with error code 6) (code=6), stdout and stderr:
> ...

**Reply by mikehardy** [reply]
I believe in Xcode you'll need to toggle on tvOS in the Info settings, but I'm not sure if that will be enough since it's build on react-native as opposed to the react-native TV fork 🤔

**Reply by mikehardy** [reply]
When I was looking at this in another module - I did a `npx react-native init ......` on the tvOS template, installed the packages I wanted to port (in this case react-native-firebase app and analytics) then fixed things directly in node_modules and copied the fixes over to the react-native-firebase test app to make sure there were no regressions, then did a PR based on that. A bit ugly + roundabout, but...it worked for me

**Reply by vladinko0** [reply]
I tried to change in Info of Xcode `tvOS Deployment` target and also in Build settings `Base SDK` to `tvOS` but it didn't help.

**Reply by vladinko0** [reply]
Looks like you are almost done with the job. I bieleve tvOS support soon will be available.
Please, do it, because I don't know how.

**Comment by prashantchothani** [comment]
Please add support for TVOS and also add to the documentation for installation in case theres anything different we need to do.

**Reply by nriccar** [reply]
Anyone can contribute, and if we require this functionality, we can help build it. I'm absolutely certain it's not ready yet because the maintainers don't have time. So, rather than demanding, try collaborating!
